File: statistics1v/SamesLib.native.py
#2345678901234567890123456789012345678901234567890123456789012345678901234567890
# WPMS2023.py - Initial prototype was  WPMS2023.rb.  See that file for full
#History.  The other language implementations will tend to stay behind that one.
#For instance, as I am having trouble with mode and quartiles in the Ruby
#verison, I will leave them out of this one for now.
#
#2345678901234567890123456789012345678901234567890123456789012345678901234567890

import csv
import re

#2345678901234567890123456789012345678901234567890123456789012345678901234567890

def isANumStr(strA):
    # TBR:  Looking for less ugly Python3 construct here, but still simple and
    # clear like in Ruby:
    if ( not isinstance(stra,str) ): 
        return False
    if ( not re.search(r'^-?\d*\.?\d+$/',strA) ):
        return False
    return True

def isNumericVector(vA):
    if ( any(isinstance(lve,Numbers.number) for lve in vA) ):
        return True
    return False

def isUsableNumber(cA):
    if ( isinstance(cA,Numbers.number) ):
        return True
    if ( isANumStr(cA) ):
        return True
    return False

def isUsableNumberVector(vA):
    if ( all(isUsableNumber(lve) for lve in vA) ):
        return True
    return False

class VectorOfX(object):

    def __init__(vectorX=[]):
        if vectorX is None:
            self.VectorOfX = []
        else:
            self.VectorOfX = vectorX

    def genCount():
        n = self.VectorOfX.size
        return n

class VectorOfContinuous(VectorOfX):

    @classmethod
    def newAfterValidation(arrayA):
        v = Array.new
        for le in arrayA:
            if not isUsableNumber(le):
                m = f"Element '{le}' should be number, this form not usable."
                raise ValueError(m)
            v.push(float(le))
        localo = self.new(v)
        return localo

    @classmethod
    def newAfterInvalidatedDropped(arrayA):
        v = Array.new
        for le in arrayA:
            if not isUsableNumber(le):
                continue
            v.push(float(le))
        localo = self.new(v)
        return localo

    def __init__(vectorX):
        super().__init__(vectorX)
        self.UseSumOfXs = False

    def assureXsPrecision(precisionSpec):
        raise ValueError("Not Yet Implemented.")

    def genInterQuartileRange():
        n = self.VectorOfX.size
                                # Subtract one here
                                # to get the offset.
        q1os    = 1                 - 1
        q2os    = ( n + 1 ) / 4     - 1
        q3os    = ( n / 2 )         - 1
        q4os    = 3 * ( q2os + 1 )  - 1
        qendos  = n                 - 1
        return q1os,  q2os,  q3os,  q4os,  qendos

    def genMax():
        svox = self.VectorOfX.sort
        return svox[-1]

    def genMean():
        n = self.VectorOfX.size.to_f
        sumxs = self.VectorOfX.sum.to_f
        return ( sumxs / n ).round(4)

    def genMeanStdDev():
        variance = nil
        if this.UseSumOfXs:
            variance = genVarianceXsSquaredMethod
        else:
            variance = genVarianceSumOfDifferencesFromMean
        stddev = Math.sqrt(variance).round(4)

    def genMedian():
        n = self.VectorOfX.size
        svox = self.VectorOfX.sort
        if ( n % 2 == 0 ):
            nm2 = ( n + 1 ) / 2
            return svox[nm2]
        else:
            nm2a = n / 2
            x1 = svox[nm2a]
            nm2b = nm2a + 1
            x2 = svox[nm2b]
            x3 = ( x1 + x2 ).to_f / 2.0
            return x3.round(4)

    def genMin():
        svox = self.VectorOfX.sort
        return svox[0]

    def genMinMax():
        svox = self.VectorOfX.sort
        return svox[0], svox[-1]

    def genMode():
        h = Hash.new
        for lx in self.VectorOfX:
            if lx in h:
                h[lx] += 1
            else:
                h[lx] = 1
        x = 0
        m = 0
        for lx in h:
            if h[lx] > m:
                x = lx
                m = h[lx]
        return x

    def genNIsEven():
        n = self.VectorOfX.size
        if ( n % 2 ) == 0:
            return True
        return False

    def genOutliers(stdDev,numberOfStdDevs=1):
        raise ValueError("Not Yet Implemented.")

    def genQuartiles():
        qos0, qos1, qos2, qos3, qos4, qose = genInterQuartileRange
        svox = self.VectorOfX.sort
        return svox[qos0], svox[qos2], svox[qos3], svox[qos4], svox[qos3]

    def genRange():
        nm1 = self.VectorOfX.size - 1
        svox = self.VectorOfX.sort
        return svox[0], svox[nm1]

    def genSum():
        sumxs = sum(self.VectorOfX)
        return sumxs

    def genVarianceSumOfDifferencesFromMean():
        mu = genMean
        n = self.VectorOfX.size
        sumofdiffsquared = 0
        for lx in self.VectorOfX:
            xlessmu = lx - mu
            sumofdiffsquared += ( xlessmu * xlessmu )
        v = sumofdiffsquared / ( n - 1 )
        return v

    def genVarianceXsSquaredMethod():
        mu = genMean
        n = self.VectorOfX.size
        sumxssquared = 0
        for lx in self.VectorOfX:
            sumxssquared += lx * lx
        v = ( sumxssquared - ( mu * mu ) ) / ( n - 1 )
        return v

    def pushX(xFloat):
        if not isUsableNumber(xFloat):
            raise ValueError
        lfn = xFloat.to_f
        self.VectorOfX.push(lfn)

    def pushX(xFloat):
        if ( not isNumber(xFloat) ):
            raise ValueError
        self.VectorOfX.push(xFloat)


class VectorOfDiscrete(VectorOfX):
    # TBD for use with columns having discrete values.

    def __init__(vectorX):
        super().__init__(vectorX)
        self.UseSumOfXs = False


class VectorTable(object):

    @classmethod
    def isAllowedDataVectorClass(vectorClass):
        if issubclass(vectorClass, Vecto):
            return True
        return False

    @classmethod
    def newFromCSV(fSpec,vcSpec,skipFirstLine=None):
        if skipFirstLine is None:
            skipFirstLine = true
            
        localo = self.new(vcSpec)
        with open(fSpec) as fp:
            i = 0
            for ll in fp:
                sll = ll.strip
                if not ( i == 0 and skipFirstLine ):
                    columns = sll.parse_csv
                    localo.pushTableRow(columns)
                i += 1
        return localo


    def __init__(vectorOfClasses):
        if not type(vectorOfClasses) is list:
            raise ValueError
        this.TableOfVectors     = Array.new
        this.VectorOfClasses    = vectorOfClasses
        for i, v in enumerate(this.VectorOfClasses):
            if lci:
                if not isAllowedDataVectorClass(lci):
                    raise ValueError 
            if lci:
                this.TableOfVectors[i] = lci.new
            else:
                this.TableOfVectors[i] = None

    def getVectorObject(indexNo):
        if not VectorTable.isAllowedDataVectorClass( this.TableOfVectors[indexNo] ):
            raise ValueError(f"Column {indexNo} not configured for Data Processing.")
        return this.TableOfVectors[indexNo]

    def pushTableRow(arrayA):
        for lvoe, buffer in zip(this.TableOfVectors,arrayA):
            if isinstance(lvoe, VectorOfX):
                lvoe.pushX(buffer)


#2345678901234567890123456789012345678901234567890123456789012345678901234567890
# End of WPMS2023.py

File: statistics1v/SamesLib.native.js
/*345678901234567890123456789012345678901234567890123456789012345678901234567890
WPMS2023.js - Initial prototype was  WPMS2023.rb.
 */

"use strict";

const FS = require('fs');

//345678901234567890123456789012345678901234567890123456789012345678901234567890

function isANumStr(strA) {
    if ( typeof strA !== 'string' ) return false;
    let lstr = strA.trim();
    if ( /^-?[0-9]*\.?[0-9]+$/.test(lstr) ) return true;
    return false;
}
module.exports.isANumStr            = isANumStr;

function isNumber(vA) {
    if ( typeof vA === 'number' ) return true;
    return false;
}
module.exports.isNumber             = isNumber;

function isNumericVector(vA) {
    if ( vA.every(isNumber) ) return true;
    return false;
}
module.exports.isNumericVector      = isNumericVector;

function isOnlyWhitespace(strA) {
    if ( typeof strA !== 'string' ) {
        throw new Error(`Invalid Argument:  Must Be String:  ${strA}.`);
    }
    if ( /^\s*$/.test(strA) ) return true;
    return false;
}
module.exports.isOnlyWhitespace     = isOnlyWhitespace;

function isUsableNumber(cA) {
    if ( isNumber(cA) )     return true;
    if ( isANumStr(cA) )    return true;
    return false
}
module.exports.isUsableNumber       = isUsableNumber;

function isUsableNumberVector(vA) {
    if ( vA.every(isUsableNumber) ) return true;
    return false;
}
module.exports.isUsableNumberVector = isUsableNumberVector;

class VectorOfX {

    _VectorOfX;

    assureXsPrecision(precisionSpec) {
        throw new Error("Not Yet Implemented.");
    }

    constructor(vectorX=Array()) {
        if ( Array.isArray( vectorX ) ) {
            this._VectorOfX = vectorX;
        } else {
            throw new Error("SyntaxError:  Argument Must Be Usable as a Number.");
        }
    }

    genCount() {
        let n = this._VectorOfX.length;
        return n;
    }

    pushX(xFloat) { throw new Error("Pure Virtual."); }

}
module.exports.VectorOfX            = VectorOfX;

class VectorOfContinuous extends VectorOfX {

    UseSumOfXs;

    constructor(vectorX=[]) {
        super(vectorX);
        this.UseSumOfXs = false;
    }

    genInterQuartileRange() {
        let n = this._VectorOfX.length;
                                // Subtract one here
                                // to get the offset.
        let q1os    = 1                 - 1;
        let q2os    = ( n + 1 ) / 4     - 1;
        let q3os    = ( n / 2 )         - 1;
        let q4os    = 3 * ( q2os + 1 )  - 1;
        let qendos  = n                 - 1;
        return [ q1os,  q2os,  q3os,  q4os,  qendos ];
    }

    genMax() {
        let svox = this._VectorOfX.sort();
        return svox[-1];
    }

    genMean() {
        let n = this._VectorOfX.length;
        let sumxs = this.genSum();
        let mu = Number( ( sumxs / n ).toFixed(4) );
        return mu;
    }

    genMeanStdDev() {
        let variance = null;
        if ( this.UseSumOfXs ) {
            variance = this.genVarianceXsSquaredMethod();
        } else {
            variance = this.genVarianceSumOfDifferencesFromMean();
        }
        let stddev = Number( Math.sqrt(variance).toFixed(4) );
        return stddev
    }

    genMedian() {
        // This is broken.  Do NOT debug until later.  TBD
        let n       = this._VectorOfX.length;
        let svox    = this._VectorOfX.sort;
        if ( n % 2 == 0 ) {
            let nm2 = ( n + 1 ) / 2;
            return svox[nm2];
        } else {
            let nm2a = n / 2;
            let x1 = svox[nm2a];
            let nm2b = nm2a + 1;
            let x2 = svox[nm2b];
            let x3 = ( x1 + x2 ) / 2.0;
            let v = Number( Math.sqrt(x3).toFixed(4) );
            return v
        }
    }

    genMin() {
        let svox = this._VectorOfX.sort()
        return svox[0]
    }

    genMinMax() {
        let svox = this._VectorOfX.sort()
        return svox[0], svox.slice(-1);
    }

    genMode() {
        // This is known to be broken in Ruby.  Need attention here too.  TBD
        let h = Hash.new
        for ( const lx of this._VectorOfX ) {
            if ( lx in h )  h[lx] += 1;
            else            h[lx] = 1;
        }
        let x = 0
        let max = 0
        for ( const lx in h ) {
            if ( h[lx] > max ) {
                let x = lx;
                let max = h[lx];
            }
        }
        return x;
    }

    genNIsEven() {
        let n = this._VectorOfX.length;
        if ( n % 2 == 0 ) return true;
        return false;
    }

    genOutliers(stdDev,numberOfStdDevs=1) {
        throw new Error("Not Yet Implemented.");
    }

    genQuartiles() {
        throw new Error("Not Yet Implemented.");
    }

    genRange() {
        let svox    = this._VectorOfX.sort
        return svox[0], svox.slice(-1);
    }

    genSum() {
        let sumxs = this._VectorOfX.reduce((partial_sum,a) => partial_sum + a, 0);
        /*
        let sumxs = 0
        for ( const lx of this._VectorOfX ) {
            sumxs += Number(lx);
        }
         */
        return sumxs
    }

    genVarianceSumOfDifferencesFromMean() {
        let mu                  = this.genMean();
        let n                   = this._VectorOfX.length;
        let sumofdiffsquared    = 0;
        for ( const lx of this._VectorOfX ) {
            let xlessmu = lx - mu;
            sumofdiffsquared += ( xlessmu * xlessmu )
        }
        let variance = sumofdiffsquared / ( n - 1 );
        return variance;
    }

    genVarianceXsSquaredMethod() {
        let mu = this.genMean();
        let n = this._VectorOfX.length;
        let sumxssquared = 0;
        for ( const lx of this._VectorOfX ) {
            sumxssquared += ( lx * lx );
        }
        let variance = ( sumxssquared - ( mu * mu ) ) / ( n - 1 );
        return variance;
    }

    pushX(xFloat) {
        let n = this._VectorOfX.length;
        if ( isNumber(xFloat) ) {
            this._VectorOfX.push(xFloat);
            n = this._VectorOfX.length;
        } else
            throw new Error(`SyntaxError:  Argument '${xFloat}', of type '${typeof xFloat}', must be usable as a number.`);
    }

}
module.exports.VectorOfContinuous   = VectorOfContinuous;

class VectorOfDiscrete extends VectorOfX {
    // TBD for use with columns having discrete values.

    constructor(vectorX=[]) {
        super();
    }

}
module.exports.VectorOfDiscrete     = VectorOfDiscrete;

class VectorTable {

    static isAllowedDataVectorClass(vectorClass) {
        if ( vectorClass === VectorOfContinuous ) return true
        return false
    }

    static newFromCSV(fSpec,vcSpec,skipFirstLine=true) {
        let localo = new VectorTable(vcSpec);
        let i = 0;

        const data = FS.readFileSync(fSpec, 'utf8');
        for ( const ll of data.split('\n') ) {
            if ( ( i == 0 ) && ( skipFirstLine ) ) {
                i += 1;
                continue
            }
            if ( isOnlyWhitespace(ll) ) {
                i += 1;
                if ( localo.IgnoreBadLines ) continue;
                throw new Error(`Blank or whitespace only line:  |${ll}|`);
            }
            let sll = ll.trim();
            let columns = sll.split(/\s*,\s*/);
            if ( localo.RowColumns != columns.length ) {
                i += 1;
                if ( this.IgnoreBadLines ) continue;
                throw new Error(`${i+1}th row had ${columns.length} columns, rather than ${localo.RowColumns}.`);
            }
            i += 1;
            if ( localo.pushTableRow(columns) ) continue;
            throw new Error(`Invalid Data Line [$i]: |${sll}|`);
        }
        return localo;
    }

    constructor(vectorOfClasses) {
        if ( vectorOfClasses === 'array' ) {
            throw new Error("SyntaxError:  Argument Must 'array'.");
        }

        this.IgnoreBadLines     = true;
        this.RowColumns         = vectorOfClasses.length;
        this.TableOfVectors     = [];
        this.VectorOfClasses    = vectorOfClasses;
        let buffer = null;
        for ( const lci of this.VectorOfClasses ) {
            if ( lci ) {
                if ( ! VectorTable.isAllowedDataVectorClass(lci) ) {
                    throw new Error("SyntaxError:  Class must be in set VectorTable accepts.");
                }
                buffer = new lci;
            } else {
                buffer = null;
            }
            this.TableOfVectors.push( buffer );
        }
    }

    getVectorObject(indexNo) {
        if ( VectorTable.isAllowedDataVectorClass( this.TableOfVectors[indexNo] ) ) {
            throw new Error(`SyntaxError:  Column ${indexNo} not configured for Data Processing.`);
        }
        return this.TableOfVectors[indexNo];
    }

    pushTableRow(arrayA) {
        let i = 0;
        for ( let lvoe of this.TableOfVectors ) {
            if ( lvoe ) { 
                let buffer = Number(arrayA[i])
                if ( isNumber(buffer) ) lvoe.pushX(buffer);
                else                    return false;
            }
            i += 1;
        }
        return true;
    }

}
module.exports.VectorTable            = VectorTable;

//345678901234567890123456789012345678901234567890123456789012345678901234567890
// End of WPMS2023.js
